### ISSUE TITLE
Request processing terminates abruptly due to async cancellation

### ISSUE THREAD

**Issue by corenti13711539** [http4s/http4s#2442]
Sometimes under high load request processing seems to terminate abruptly with the following message logged by `ResponseLogger` middleware:
```service cancelled response```

This seems to result in no response getting sent to the client e.g.:
```
xyz:load-test foo$ curl -v http://localhost:8080/api/foo/op1
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET /api/foo/op1 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
```

Seems to be related to blaze server as after switching to Jetty the issue seems to have disappeared.

**Comment by corenti13711539** [comment]
code for reproducing the issue:
[http4s-demo.zip](https://github.com/http4s/http4s/files/2925730/http4s-demo.zip)

**Comment by corenti13711539** [comment]
@rossabaker @ChristopherDavenport Referring to Gitter discussions on async cancellation failures with http4s 0.20, I think I was able to create a service implementation for reproducing the issue. Could you please check out the project to see if you're able to reproduce the issue. Thanks!

**Comment by ChristopherDavenport** [comment]
@corenti13711539 Can you place that either in a repo or gist please? A zip is by default a worrisome form of transpot.

**Comment by corenti13711539** [comment]
@ChristopherDavenport Just created a repo for this: https://github.com/corenti13711539/http4s-async-cancel

**Comment by rossabaker** [comment]
Thanks for your patience.  I finally had a chance to study this.  I submitted some suggestions at https://github.com/corenti13711539/http4s-async-cancel/pull/1.  The global execution context is blocked on both `MainServer` and `SlowServer`, leaving it unable to respond.  With the suggested changes, we still simulate blocking APIs, but the server remains responsive.

**Comment by corenti13711539** [comment]
Thanks a lot for taking the time to study this issue and the PR @rossabaker ! 👍🙇‍

Still wondering if there could be more to this, though. Here's some more background.

`SlowServer` tries to simulate a service that's slow to respond. Unfortunately the service is not controlled by us and we're unable to do any changes. So, for that reason I applied the changes from your PR in `MainServer`, but purposely left `SlowServer` unchanged. AFAIK, this should prevent `MainServer` from blocking on global execution context, but I still get lots of messages like the following in `MainServer` (with the changes applied) when running the load test:
```
2019-03-31 11:07:08.301 [scala-execution-context-global-26] INFO  org.http4s.server.middleware.Logger - service cancelled response
```

In my actual application I'm using monadic combinators similar to this:
https://github.com/corenti13711539/http4s-async-cancel/blob/master/src/main/scala/foo/MainServer.scala#L119-L122

Sometimes request processing seems to just stop abruptly without any trace of errors or a response getting delivered. I've tried to trace the exact location and this seems to occur in the middle of a for comprehension similar to that above.

Is it possible that http4s-blaze would occasionally drop accepted requests in code like the above in
high-load scenarios?

**Comment by rossabaker** [comment]
Okay, I can still reproduce the ESOCKETTIMEDOUT in artillery when SlowServer blocks threads as opposed to just responding too slowly.

**Comment by rossabaker** [comment]
In blaze-client, it fails with a `WaitQueueFullFailure`.  async-http-client is failing with `TooManyConnectionsPerHost`.  But I would still expect a 500 error from the MainServer.

**Comment by rossabaker** [comment]
It has nothing to do with the client.  It can be simulated with `Async[F].never`

**Comment by rossabaker** [comment]
blaze-server responds with a 503 if the response header timeout (default 1 minute) fires.  You can see this behavior by adding `.withResponseHeaderTimeout(10.seconds)` to the `BlazeServerBuilder`.

If the idle timeout (default 30 seconds) fires, the connection is [closed without a response](https://github.com/http4s/http4s/blob/949d028b11b5ad41c8ea577135fd511decabb68a/blaze-server/src/main/scala/org/http4s/server/blaze/Http1ServerStage.scala#L118).  The idle timeout does not know whether we've already responded or not.  To get a 503 in this case, we would need to know at that point whether the response is "committed" by having already sent a header.  This would be friendlier, but might also be expensive.

**Comment by rossabaker** [comment]
The fix proposed in the last comment is at #2550, though it can be worked around now.  Please reopen if it's still a problem.
